src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,403 @@
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	int16_t *dst = (int16_t *)sink->w_ptr + start_frame;
+	int16_t *src = (int16_t *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = sat_int16((int32_t)*dst + (int32_t)*src++);
+			dst++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++)
+			*dst++ = *src++;
+	}
+}
+
+static void remap_mix_channel_s16(struct audio_stream __sparse_cache *sink,
+				  uint8_t sink_channel_index,  uint8_t sink_channel_count,
+				  uint32_t start_frame, uint32_t mixed_frames,
+				  const struct audio_stream __sparse_cache *source,
+				  uint8_t source_channel_index, uint8_t source_channel_count,
+				  uint32_t frame_count, uint16_t gain)
+{
+	int16_t *dst, *src;
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, frames, i;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	dst = (int16_t *)sink->w_ptr + start_frame * sink_channel_count + sink_channel_index;
+	src = (int16_t *)source->r_ptr + source_channel_index;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= frames) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames * source_channel_count, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		/* frames is the processed frame count in this loop*/
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			*dst = sat_int16((int32_t)*dst +
+			       q_mults_16x16(*src, gain, IPC4_MIXIN_GAIN_SHIFT));
+			src += source_channel_count;
+			dst += sink_channel_count;
+			frames++;
```

Comment:
still needlessly counting `frames` - here and in other locations

Reply:
you think a single division is more expensive than possibly hundreds of increments? But ok, not really important.

Reply:
actually, it is because the author of original functions comments he doesn't  want a division, so I avoid to use division here.

Reply:
normally, divide(int) takes up to 32 cycles in dsp, and divide also can be done with >> partially.
The code looks complex, still because the wrapper buffer usage